common/src/main/java/io/crate/types/RowType.java

```diff
@@ -0,0 +1,162 @@
+            }
+            this.fieldNames = List.copyOf(generatedFieldNames);
+        } else {
+            this.fieldNames = List.copyOf(fieldNames);
+        }
+    }
+
+    public RowType(List<DataType<?>> fieldTypes) {
+        this(fieldTypes, List.of());
+    }
+
+    public RowType(StreamInput in) throws IOException {
+        int numFields = in.readVInt();
+        ArrayList<DataType<?>> fieldTypes = new ArrayList<>(numFields);
+        ArrayList<String> fieldNames = new ArrayList<>(numFields);
+        for (int i = 0; i < numFields; i++) {
+            fieldTypes.add(DataTypes.fromStream(in));
+            fieldNames.add(in.readString());
+        }
+        this.fieldTypes = List.copyOf(fieldTypes);
+        this.fieldNames = List.copyOf(fieldNames);
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(fieldTypes.size());
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            DataTypes.toStream(fieldTypes.get(i), out);
+            out.writeString(fieldNames.get(i));
+        }
+    }
+
+    public int numElements() {
+        return fieldTypes.size();
+    }
+
+    public DataType<?> getFieldType(int position) {
+        return fieldTypes.get(position);
+    }
+
+    public String getFieldName(int position) {
+        return fieldNames.get(position);
+    }
+
+    public List<String> fieldNames() {
+        return fieldNames;
+    }
+
+    public List<DataType<?>> fieldTypes() {
+        return fieldTypes;
+    }
+
+    @Override
+    public Row readValueFrom(StreamInput in) throws IOException {
+        Object[] values = new Object[fieldTypes.size()];
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            values[i] = fieldTypes.get(i).streamer().readValueFrom(in);
+        }
+        return new RowN(values);
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public void writeValueTo(StreamOutput out, Row row) throws IOException {
+        assert row.numColumns() == fieldTypes.size()
+            : "Row that should be streamed must have the same number of columns as the rowType contains fieldTypes";
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            Streamer streamer = fieldTypes.get(i).streamer();
+            streamer.writeValueTo(out, row.get(i));
+        }
+    }
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public Precedence precedence() {
+        return Precedence.TableType;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public Streamer<Row> streamer() {
+        return this;
+    }
+
+    @Override
+    public Row value(Object value) throws IllegalArgumentException, ClassCastException {
+        return (Row) value;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public int compareValueTo(Row val1, Row val2) {
+        for (int i = 0; i < fieldTypes.size(); i++) {
```

Comment:
not sure, maybe we can first`Row#numColumns`  of val1 and val2?

Reply:
I added an assertion to make sure they have the same size. I think comparing different rows to each other shouldn't be allowed. (This here is for sorting, and we can't sort different row values).